File: 프로그래머스/2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.py
def solution(files):
    answer = []
    sep = []
    for file in files:
        head, number, tail = '','',''
        for f in file:
            if not head or (not f.isnumeric() and not number):
                head += f
            elif f.isnumeric() and head and not tail:
                number += f
            elif tail or (number and not f.isnumeric()):
                tail += f
        sep.append([head, number, tail])
    
    sep.sort(key=lambda x:(x[0].lower(), int(x[1])))
    for s in sep:
        answer.append(''.join(s))
    return answer

File: 프로그래머스/2/17686. ［3차］ 파일명 정렬/［3차］ 파일명 정렬.js
function solution(files) {
    let answer = [];
    let sepFiles = [];
    
    for(let i=0; i<files.length; i++) {
        let head = '', number='', tail='';
        for (let j=0; j<files[i].length; j++) {
            const f = files[i][j];
            if (head.length==00 || ((isNaN(f) || f===' ') && number.length===0)) head += f;
            else if (!isNaN(f) && head.length>0 && tail==='') number += f;
            else if (tail.length>0 || (number.length>0 && (isNaN(f) || f===' '))) tail += f;
        }
        sepFiles.push([head, number, tail]);
    }
    sepFiles
        .sort((a, b) => (a[0].toLowerCase().localeCompare(b[0].toLowerCase())) || Number(a[1]) - Number(b[1]))
        .forEach((file) => answer.push(file.join("")));
    
    return answer;
}